tool/planet/agent.go

```diff
@@ -173,34 +179,38 @@ func startLeaderClient(conf *LeaderConfig, errorC chan error) (leaderClient io.C
 		enabled, _ := strconv.ParseBool(newVal)
 		switch enabled {
 		case true:
-			if cancelVoter != nil {
-				log.Infof("voter is already active")
+			if voterCancel != nil {
+				logger.Info("Voter is already active.")
 				return
 			}
-			// start election participation
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
-				log.Errorf("failed to add voter for %v: %v", conf.PublicIP, trace.DebugReport(err))
+			logger.Info("Start election participation.")
+			voterCtx, voterCancel = context.WithCancel(context.TODO())
+			if err = client.AddVoter(voterCtx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
+				voterCancel()
+				log.WithError(err).Error("Failed to add voter.")
 				errorC <- err
 			}
 		case false:
-			if cancelVoter == nil {
-				log.Info("no voter active")
+			if voterCancel == nil {
+				log.Info("No voter active.")
 				return
 			}
-			// stop election participation
-			cancelVoter()
-			cancelVoter = nil
+			logger.Info("Stop election participation.")
+			voterCancel()
+			voterCancel = nil
 		}
 	})
 	// modify /etc/hosts upon election of a new leader node
 	client.AddWatchCallback(conf.LeaderKey, conf.Term/3, func(key, prevVal, newLeaderIP string) {
-		if err := updateDNS(conf, hostname, newLeaderIP); err != nil {
-			log.Error(trace.DebugReport(err))
+		if err := updateDNS(conf, newLeaderIP); err != nil {
+			log.WithError(err).WithField("leader-addr", newLeaderIP).Error("Failed to update DNS.")
```

Comment:
Not really related to your change, but I wonder if the way we handle these errors (here when updating DNS and above when stopping/starting services) can be potentially causing some issues with our leader election. For example, if we failed to update it here, shouldn't we keep retrying? Otherwise this node will be talking to a wrong apiserver until the next leader change.